samcli/hook_packages/terraform/hooks/prepare/types.py

```diff
@@ -97,3 +98,69 @@ def validate(self):
         the AWS SAM CLI transformation logic for the given resource
         """
         raise NotImplementedError
+
+
+@dataclass
+class ResourceTranslationProperties:
+    """
+    These are a collection of properties useful for defining a
+    translation from Terraform to CFN for a given resource
+    """
+
+    resource: Dict
+    translated_resource: Dict
+    config_resource: TFResource
+    logical_id: str
+    resource_full_address: str
+
+
+class ResourceProperties(ABC):
+    """
+    Interface definition for a resource type to handle collecting and storing properties specific to it's type.
+    These properties are then used for handling specific resource-based logic, such as linking to other resources.
+    """
+
+    def __init__(self):
+        self.terraform_resources: Dict[str, Dict] = {}
+        self.terraform_config: Dict[str, TFResource] = {}
+        self.cfn_resources: Dict[str, List] = {}
+
+    def collect(self, properties: ResourceTranslationProperties):
+        """
+        Collect any properties required for future resource processing after traversing the Terraform modules.
+
+        Parameters
+        ----------
+        properties: ResourceTranslationProperties
+            Properties acquired specific to an AWS resource when iterating through a Terraform module
```

Comment:
The reason it isn't an instance variable is because the class is instantiated once, but the collect method can be called multiple times for each resource of that type.